Add a new metric for some minor SRU statistics.

sil2100 opened the pull request:
For now only two - the SRU queue length at a given moment (not counting Backports and partner) and how many ready-to-release SRUs are in -proposed.

sil2100 commented:
If anyone is wondering: I'm parsing pending-srus instead of doing LP calls as it's usually just easier to do that, if we're not querying too often. Brian's scripts also do this generally. I can switch to using LP API if anyone finds that better.

OddBloke commented on metrics/foundations_sru.py:
```diff
@@ -0,0 +1,125 @@
+
+try:
+    from html.parser import HTMLParseError
+except ImportError as exception:
```
When might we hit this?

sil2100 replied:
I have no idea what machine is the metrics stuff running on, but from what I saw HTMLParseError is deprecated from python >= 3.5. Strangeness.

OddBloke commented:
I'd like to understand that import thing, but other than that this looks good. :smile:

murraybd commented:
I was already working on this myself and would prefer to have a chance to review it before it gets merged.

sil2100 commented:
I wasn't aware Brian already had this on his plate. Feel free to drop this if it's duplicated work, didn't want to step on anyone's toes...

murraybd commented on metrics/foundations_sru.py:
```diff
@@ -0,0 +1,125 @@
+
+
+def sru_queue_count():
+    """Get the number of UNAPPROVED uploads for each series."""
```
I think we should be clear here and indicate that the count is of unapproved and proposed.

murraybd commented on metrics/foundations_sru.py:
```diff
@@ -0,0 +1,125 @@
+        per_series[series.name] = len(series.getPackageUploads(
+            status='Unapproved',
+            pocket='Proposed',
+            archive=ubuntu.main_archive))
```
We should include partner because that is something the SRU team would review, however not including backports i correct because there is another team for that.

murraybd commented on metrics/foundations_sru.py:
```diff
@@ -0,0 +1,125 @@
+
+def sru_ready_for_updates_count():
+    """Get the number of verified -proposed packages."""
+    # Most of this code is taken from lp:~brian-murray/+junk/bug-agent, just
```
The function name and the docstring differ here in that not everything which is verified is ready for release.  We would not release a package which has failing autopkgtests. So what are we really counting here? The quantity of things verified or the quantity of things that are releasable? I'd argue for that latter since that represents work for the SRU team and as such we should exclude things that FTBFS, have a depwait, or an autopkgtest regression. All of those are things for the uploader to address.

murraybd commented on metrics/foundations_sru.py:
```diff
@@ -0,0 +1,125 @@
+            cols = tag.findAll('td')
+            length = len(cols)
+            if length == 0:
+                continue
```
Around here I'd add the check for depwait, FTBFS, and regressions - something like this should be close:

        failure = cols[0].text.replace(package, '') 
        if 'Failed' in failure or 'Dependency wait' in failure or 'Cancelled' in failure:
            continue
        if 'Regression in autopkgtest' in failure:
            continue

murraybd commented on metrics/foundations_sru.py:
```diff
@@ -0,0 +1,125 @@
+def collect(dryrun=False):
+    """Collect and push SRU-related metrics."""
+    sru_queues = sru_queue_count()
+    ready_srus = sru_ready_for_updates_count()
```
Again, I'd consider renaming this to more accurately reflect what it really is.

murraybd commented on metrics/foundations_sru.py:
```diff
@@ -0,0 +1,125 @@
+        registry = CollectorRegistry()
+
+        gauge = Gauge('foundations_sru_unapproved_count',
+                      'Number of Uploads in the Unapproved Queue per Series',
```
I think this should be renamed so that it also includes proposed, so foundations_sru_unapproved_proposed_count. Additionally, you might make the string which is also used in the print a variable.

sil2100 commented:
Thanks for the review! Hopefully I addressed most of your concerns here - but please be aware that function names need to have 30 or less characters so usually I just put naming that 'approximately' fits the function purpose. It's hard to find a good name when 30 characters is the limit (per pylint). Also, I suck in naming things, so to not waste time by bikesheding things by myself for 30 minutes I just take the closest bet and use that. Propositions are always welcome.

murraybd commented on metrics/foundations_sru.py:
```diff
@@ -0,0 +1,131 @@
+
+        gauge = Gauge(
+            'foundations_sru_unapproved_proposed_count',
+            'Number of Proposed Uploads in the Unapproved Queue per Series',
```
Given other discussions about this metric not being specific to the foundations team, it is useful to people outside the foundations team, I think it should be renamed to "distro_sru_unapproved_proposed_count". I prefer distro because it feels like more of a community term than platform which was another suggestion.

murraybd commented on metrics/foundations_sru.py:
```diff
@@ -0,0 +1,131 @@
+            gauge.labels(series).set(count)
+
+        gauge = Gauge(
+            'foundations_sru_verified_and_ready_count',
```
I'd replace foundations_ with distro_ here too.

sil2100 commented:
Renamed as per review. I didn't change the name of the python file though, since I guess it's nice to have a good distinction which of the metrics are maintained by us and which by the server team.
